gps with nil pointer dereference

<!--

Thanks for filing an issue! If this is a question or feature request, just delete
everything here and write out the request, providing as much context as you can.

-->

### What version of `dep` are you using (`dep version`)?
<!--
  If you installed `dep` via `go get`, report the version instead with
  `cd $GOPATH/src/github.com/golang/dep && git describe --tags`
-->
 version     : devel
 build date  : 
 git hash    : 
 go version  : go1.9
 go compiler : gc
 platform    : linux/amd64

### What `dep` command did you run?

dep ensure -update
<!--

Paste the output of the commands you ran in here, making sure to pass -v for maximum context.

The output of `dep hash-inputs` may also be helpful to include.

-->

```
Root project is "bitbucket.org/arquivei/go-dfeadd"
 13 transitively valid internal packages
 28 external packages imported from 13 projects
(0)   ✓ select (root)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0x755c10]

goroutine 1 [running]:
github.com/golang/dep/internal/gps.(*sourceGateway).sourceURL(0xc4200624e0, 0xb129c0, 0xc420016278, 0x0, 0x0, 0x0, 0x0)
	/home/victor/go/src/github.com/golang/dep/internal/gps/source.go:479 +0x100
github.com/golang/dep/internal/gps.(*sourceCoordinator).getSourceGatewayFor(0xc420322540, 0xb129c0, 0xc420016278, 0xc42001d470, 0x23, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/victor/go/src/github.com/golang/dep/internal/gps/source.go:161 +0x975
github.com/golang/dep/internal/gps.(*SourceMgr).SourceExists(0xc4202f41e0, 0xc42001d470, 0x23, 0x0, 0x0, 0xc420028000, 0x0, 0x40)
	/home/victor/go/src/github.com/golang/dep/internal/gps/source_manager.go:467 +0x91
github.com/golang/dep/internal/gps.(*bridge).SourceExists(0xc4201da510, 0xc42001d470, 0x23, 0x0, 0x0, 0xc42012ac38, 0xc42012acd8, 0x54011b)
	/home/victor/go/src/github.com/golang/dep/internal/gps/bridge.go:132 +0x93
github.com/golang/dep/internal/gps.(*solver).createVersionQueue(0xc420144200, 0xc42001d470, 0x23, 0x0, 0x0, 0xc42022cf80, 0x8, 0x8, 0x0, 0x0, ...)
	/home/victor/go/src/github.com/golang/dep/internal/gps/solver.go:767 +0xed
github.com/golang/dep/internal/gps.(*solver).solve(0xc420144200, 0x0, 0x0, 0xc4201da3f0)
	/home/victor/go/src/github.com/golang/dep/internal/gps/solver.go:492 +0x4cf
github.com/golang/dep/internal/gps.(*solver).Solve(0xc420144200, 0x34, 0xb10440, 0xb6a488, 0xc42001e0d8)
	/home/victor/go/src/github.com/golang/dep/internal/gps/solver.go:442 +0xaa
main.(*ensureCommand).runUpdate(0xc4201031e0, 0xc420088fa0, 0xc4200100f0, 0x0, 0x0, 0xc420011440, 0xb15480, 0xc4202f41e0, 0xc42001e0c4, 0x34, ...)
	/home/victor/go/src/github.com/golang/dep/cmd/dep/ensure.go:383 +0x8c1
main.(*ensureCommand).Run(0xc4201031e0, 0xc420088fa0, 0xc4200100f0, 0x0, 0x0, 0x0, 0x0)
	/home/victor/go/src/github.com/golang/dep/cmd/dep/ensure.go:188 +0x3ec
main.(*Config).Run(0xc420062f00, 0xc420062f00)
	/home/victor/go/src/github.com/golang/dep/cmd/dep/main.go:159 +0x7a2
main.main()
	/home/victor/go/src/github.com/golang/dep/cmd/dep/main.go:45 +0x1fc

```

### What did you expect to see?

### What did you see instead?

--- a/internal/gps/source_errors.go
+++ b/internal/gps/source_errors.go
@@ -13,14 +13,23 @@ import (
 // preserving the actual vcs command output and error, in addition to the message.
 // All other types pass through unchanged.
 func unwrapVcsErr(err error) error {
+	var cause error
+	var out, msg string
+
 	switch t := err.(type) {
 	case *vcs.LocalError:
-		cause, out, msg := t.Original(), t.Out(), t.Error()
-		return errors.Wrap(errors.Wrap(cause, out), msg)
+		cause, out, msg = t.Original(), t.Out(), t.Error()
 	case *vcs.RemoteError:
-		cause, out, msg := t.Original(), t.Out(), t.Error()
-		return errors.Wrap(errors.Wrap(cause, out), msg)
+		cause, out, msg = t.Original(), t.Out(), t.Error()
+
 	default:
 		return err
 	}
+
+	if cause == nil {
+		cause = errors.New(out)
+	} else {
+		cause = errors.Wrap(cause, out)
+	}
+	return errors.Wrap(cause, msg)
 }

--- a/internal/gps/source_errors_test.go
+++ b/internal/gps/source_errors_test.go
@@ -1,0 +1,30 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gps
+
+import (
+	"testing"
+
+	"github.com/Masterminds/vcs"
+)
+
+func TestUnwrapVcsErrNonNil(t *testing.T) {
+	for _, err := range []error{
+		vcs.NewRemoteError("msg", nil, "out"),
+		vcs.NewRemoteError("msg", nil, ""),
+		vcs.NewRemoteError("", nil, "out"),
+		vcs.NewRemoteError("", nil, ""),
+		vcs.NewLocalError("msg", nil, "out"),
+		vcs.NewLocalError("msg", nil, ""),
+		vcs.NewLocalError("", nil, "out"),
+		vcs.NewLocalError("", nil, ""),
+		&vcs.RemoteError{},
+		&vcs.LocalError{},
+	} {
+		if unwrapVcsErr(err) == nil {
+			t.Errorf("unexpected nil error unwrapping: %#v", err)
+		}
+	}
+}

